uReplicator-Manager/src/main/java/com/uber/stream/kafka/mirrormaker/manager/core/ControllerHelixManager.java

```diff
@@ -1089,47 +1092,42 @@ public InstanceTopicPartitionHolder createNewRoute(String pipeline, int routeId)
     }
 
     _availableControllerList.remove(instanceName);
-    _pipelineToInstanceMap.put(pipeline, new PriorityQueue<>(1,
+    _pipelineToInstanceMap.putIfAbsent(pipeline, new PriorityQueue<>(1,
         InstanceTopicPartitionHolder.totalWorkloadComparator(_pipelineWorkloadMap)));
     _pipelineToInstanceMap.get(pipeline).add(instance);
     _assignedControllerCount.inc();
     _workerHelixManager.addTopicToMirrorMaker(instance, pipeline, routeId);
 
     // register metrics
-    String routeString = srcDst[1] + "-" + srcDst[2] + "-" + routeId;
     maybeRegisterMetrics(routeString);
 
     spectator.disconnect();
     return instance;
   }
 
   public InstanceTopicPartitionHolder maybeCreateNewRoute(
-      PriorityQueue<InstanceTopicPartitionHolder> instanceList,
-      String topicName,
-      int numPartitions,
-      String pipeline) throws Exception {
+          PriorityQueue<InstanceTopicPartitionHolder> instanceList,
+          String topicName,
+          int numPartitions,
+          String pipeline) throws Exception {
 
     LOGGER.info("maybeCreateNewRoute, topicName: {}, numPartitions: {}, pipeline: {}", topicName, numPartitions,
-        pipeline);
+            pipeline);
 
-    Set<Integer> routeIdSet = new HashSet<>();
     for (InstanceTopicPartitionHolder instance : instanceList) {
       if (instance.getTotalNumPartitions() + numPartitions < _initMaxNumPartitionsPerRoute) {
+        _workerHelixManager.addWorkersToMirrorMaker(pipeline, instance.getRoute().getPartition());
         return instance;
       }
-      routeIdSet.add(instance.getRoute().getPartition());
     }
 
-    // For now we don't delete route even it's empty. so routeId should be 0,1...N
     int routeId = 0;
-    while (routeId < routeIdSet.size() + 1) {
-      if (!routeIdSet.contains(routeId)) {
-        break;
+    for (InstanceTopicPartitionHolder instance : instanceList) {
+      if(routeId < instance.getRoute().getPartition()){
```

Comment:
The original logic is: if current route id is: 0, 1, 3, 4, then we create 2. Any reason you want to change that?